OpenRA.Mods.Common/Widgets/Logic/ColorPickerLogic.cs

```diff
@@ -65,6 +73,174 @@ public ColorPickerLogic(Widget widget, ModData modData, World world, HSLColor in
+					var newSwatch = (ColorBlockWidget)template.Clone();
+					newSwatch.Visible = true;
+					newSwatch.IsVisible = () => true;
+					newSwatch.Id = nameString;
+					newSwatch.X = x.ToString();
+					newSwatch.Y = y.ToString();
+					newSwatch.Initialize(new WidgetArgs());
+
+					int thisIndex = i;
+					var thisHexColor = colorList[i];
+
+					Color thisColor;
+					if (HSLColor.TryParseRGB(thisHexColor, out thisColor))
+						newSwatch.GetColor = () => thisColor;
+					else
+						newSwatch.GetColor = () => Color.White;
+
+					newSwatch.OnMouseUp = (m) =>
+					{
+						if (isEditable && Game.GetModifierKeys().HasModifier(Modifiers.Ctrl))
+						{
+							var playerSettings = Game.Settings.Player;
+
+							var newColor = mixer.Color;
+							if (thisIndex < playerSettings.CustomColors.Length)
+							{
+								playerSettings.CustomColors[thisIndex] = newColor.ToHexString();
+								var newColorRgb = Color.FromArgb(mixer.Color.RGB.ToArgb());
+								newSwatch.GetColor = () => newColorRgb;
+							}
+
+							Game.Settings.Save();
+						}
+						else
+						{
+							var newColor = newSwatch.GetColor();
+							mixer.Set(new HSLColor(newColor));
+							onChange(mixer.Color);
+						}
+					};
+
+					containerWidget.AddChild(newSwatch);
+
+					x += swatchSize + swatchPadding;
+					columnIndex++;
+					if (columnIndex >= maxColumns)
+					{
+						x = initX;
+						y += swatchSize + swatchPadding;
+						columnIndex = 0;
+					}
+				}
+			});
+
+			// Action to remove all the custom color swatches
+			var removeCustomColorSwatchesAction = new Action(() =>
+			{
+				var childWidgets = customArea.Children.Where(x => x.Visible && !string.IsNullOrEmpty(x.Id) && x.Id.StartsWith("COLORCUSTOM")).ToList();
+				foreach (var childWidget in childWidgets)
+				{
+					customArea.RemoveChild(childWidget);
+				}
+			});
+
+			// Get preset and custom colors
+			var presetColors = validator.TeamColorPresets;
+			if (presetColors == null)
+				presetColors = new string[] { };
+
+			var customColors = Game.Settings.Player.CustomColors;
+			if (customColors == null)
+				customColors = new string[] { };
+
+			// Police maximum number of custom colors
+			if (customColors.Length > maxCustomColors)
+				customColors = customColors.Take(maxCustomColors).ToArray();
+
+			// Layout swatches
+			addSwatchesAction(presetColors, presetArea, presetColorTemplate, false);
+			addSwatchesAction(customColors, customArea, customColorTemplate, true);
+
+			// Store color button
+			var storeButton = widget.GetOrNull<ButtonWidget>("STORE_BUTTON");
+			if (storeButton != null)
+			{
+				storeButton.OnClick = () =>
+				{
+					var list = customColors.ToList();
+
+					var newColor = mixer.Color;
+					var hexColor = newColor.ToHexString();
+
+					if (list.Contains(hexColor))
+						return;
+
+					if (list.Count < maxCustomColors)
+					{
+						list.Add(hexColor);
+					}
+					else
+					{
+						// Wrap around if reached maximum
```

Comment:
IMO it would feel much better ingame if this always inserted the color at the start (`list.Insert(0, hexColor)`), rolling all the older colors one step forward in the list.